cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractTwillProgramController.java

```diff
@@ -80,29 +80,32 @@ public void run() {
         if (stopRequested) {
           // Service was killed
           stop();
-        } else {
+          return;
+        }
+
+        ServiceController.TerminationStatus terminationStatus = twillController.getTerminationStatus();
+        // The terminationStatus shouldn't be null when the twillController state is terminated
+        // In case it does (e.g. bug or twill changes), rely on the termination future to determine the state
+        if (terminationStatus == null) {
           try {
-            // This never blocks since the twill controller is already terminated. It will throw exception if
-            // the twill program failed.
-            twillController.awaitTerminated();
-            // Service completed by itself. Simply signal the state change of this controller.
-            // TODO (CDAP-6806): this should not be done with reflection but through a proper Twill API
-            // Figure out whether the final Yarn status is in error, if so, set state accordingly
-            if (twillController instanceof YarnTwillController) {
-              FinalApplicationStatus finalStatus = ((YarnTwillController) twillController).getTerminationStatus();
-              if (FinalApplicationStatus.FAILED.equals(finalStatus)) {
-                complete(State.ERROR);
-                return;
-              } else if (FinalApplicationStatus.KILLED.equals(finalStatus)) {
-                complete(State.KILLED);
-                return;
-              }
-            }
-            // normal termination
+            Futures.getUnchecked(twillController.terminate());
             complete();
           } catch (Exception e) {
-            error(e);
+            complete(State.ERROR);
           }
+          return;
+        }
+        // Based on the terminationStatus to set the completion state of the program controller.
+        switch (terminationStatus) {
+          case SUCCEEDED:
+            complete();
+            break;
+          case FAILED:
+            complete(State.ERROR);
+            break;
+          case KILLED:
+            complete(State.KILLED);
+            break;
```

Comment:
Do you think its better to add the default case where we throw IllegalStateException?